src/repl/session.rs

```diff
@@ -625,6 +627,26 @@ impl Session {
         }
     }
 
+    #[cfg(feature = "cli")]
+    pub fn read(&mut self, output: &mut Vec<String>, cmd: &str) {
+        let filename = match cmd.split_once(" ") {
+            Some((_, filename)) => filename,
+            _ => return output.push(red!("Usage: ::read <filename>")),
+        };
+
+        let snippet = match fs::read_to_string(filename) {
+            Ok(snippet) => snippet,
+            Err(err) => return output.push(red!(format!("unable to read {}: {}", filename, err))),
+        };
+
+        let mut result = match self.formatted_interpretation(snippet, None, false, None, None) {
+            Ok((output, _)) => output,
+            Err(result) => result,
+        };
```

Comment:
After reading from the file, I think it should be treated the same as when a snippet is passed in. The missing piece here then would be handling when `show_costs` mode is enabled.

Perhaps a clean solution would be to refactor the `snippet => {...}` case of the match in `handle_command` into a function, then call that function here.

Reply:
I've pushed a commit for this.  It's a clean change; I was able to merge both `handle_command` and the cli `get_costs` but not the non-cli `get_costs`.  The main change in behavior from my first commit is that `::read` can now show the execution costs if the user has toggled cost reporting.

Reply:
See new review comment.